### ISSUE TITLE
How to change T2T configuration?

### ISSUE THREAD

**Issue by mzeidhassan** [ModernMT/MMT#398]
Hi MMT team,

First, thanks for adding the Transformer (T2T).  It's a very welcome addition. I didn't have time to use it yet, but I am planning to, but since this will be an 'expensive' training, I would like to figure best configuration first. 

As you may be aware, there are different configurations for T2T which are basically hyper-parameters pre-defined in T2T. For exmaple: transformer_big_single_gpu (BIG) and transformer_base_single_gpu (BASE) `--hparams_set=transformer_big_single_gpu`.

As far as I can tell, the current configuration is based on the [Base Single GPU](https://github.com/ModernMT/MMT/blob/827e95971a7d96831e57cc995bb4dda1c30a9eb1/cli/mmt/engine.py) configuration, right?

My main question is:

If I want to change the currently set hyper-parameters, where should I change that? 
I can find some parameters in MMT/cli/mmt/engine.py

I couldn't locate the learning rate parameter in this file though, but I found it in [transformer.py](https://github.com/ModernMT/MMT/blob/2c0eace2307cb981e00df584a7fa8be4d87075ef/src/decoder-neural/src/main/python/nmmt/transformer.py)

Can I simply run something like the following in the terminal:

```
./mmt create en it examples/data/train 
  --train-steps 6000
  --validation-steps 50000 
  --hparams_set=transformer_big_single_gpu
  --hparams='batch_size=2048'
  --validation-steps 50000 
  --checkpoint-steps 50000 
  --gpus 0
```

One more question:

Are we still able to use the [Advanced Configurations](https://github.com/ModernMT/MMT/wiki/Advanced-Configurations) as we used to do with 2.5 for example or it's not valid any longer? 

Thanks in advance for your support!

Kind regards,
Mohamed

**Comment by nicolabertoldi** [comment]
``mmt create`` expose only some hyper-parameters of the transformer model.

As you can see in the help
```
./mmt create --help
```
You can choose:
- the size of the transformer using the parameter ``--hparams`` which can get values among the models registered in T2T; usually we use either ``transformer_base`` (default) or ``transformer_big``.
- the training parameters (``--train-steps``, ``--validation-steps``, ``--checkpoint-steps``)
- the batch size of the training (``--batch-size``)
- the size of the vocabulary of subtokens (``-bpe``)

Other hyper-parameters of the transformer (like the learning rate) are not exposed (at the moment) in the command line. If you were interested to do it, you should edit the 

MMT uses the ``Adam`` optimizer, and its default.


The right command corresponding to yours is
```
./mmt create en it examples/data/train 
  --hparams=transformer_big
  --batch_size 2048
  --train-steps 6000 
  --validation-steps 50000 
  --checkpoint-steps 50000 
  --gpus 0
```
I would like to highlight that it is not worth to have validation-steps and checkpoint-steps larger than train-steps, because validation and saving of checkpoint cannot be performed after the training is finished.


Actually, the [Advanced Configurations](https://github.com/ModernMT/MMT/wiki/Advanced-Configurations) refers to the old version. We will update them asap. 
However, the changes are minimal:
- in ``engine`` the attribute ``type`` is no more valid, because we do not support anymore the PB model
- in ``engine`` the attributes ``source-language`` and ``target-language`` are no more valid; instead you have to specify languages through the section ``languages`` like in a multilingual engine

**Comment by mzeidhassan** [comment]
Thanks a million @nicolabertoldi for such detailed answer. It is very helpful indeed. 

It seems there is something missing in your answer regarding learning rate above after this line:

"If you were interested to do it, you should edit the"

Can you please double check?

**Comment by nicolabertoldi** [comment]
Other hyper-parameters of the transformer (like the learning rate) are not exposed (at the moment) in the command line. If you were interested to do it, you should edit the file ``cli/mmt/engine.py``
and change the line
```
hparams_p = 'batch_size=%d' % batch_size
```

This line set the parameters which overwrite the default of the chosen transformer problem.
For instance, if you want to use change the optimizer and use SGD, you can write
```
hparams_p = 'batch_size=%d,optimizer=SGD' % batch_size
```


Obviously, you have to respect syntax, parameters and values of the transformer.

**Comment by mzeidhassan** [comment]
Thanks again @nicolabertoldi ! I am grateful to your detailed and informative answers.  You may close this issue if you wish.
